### PR TITLE
Use IMDSv2

### PR DESCRIPTION
**Pull request by pb8o**
IMDSv1 has problems and nowadays IMDSv2 is preferred.

Signed-off-by: Pablo Barbáchano <pablob@amazon.com>

# Reason for This PR

`[Author TODO: add issue #.]`
`[Open the PR after the related issue has a clear conclusion.]`
`[If there is no issue which states the need for this PR, create one first.]`
Fixes #

## Description of Changes

`[Author TODO: add description of changes.]`

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes follow the [Runbook for Firecracker API changes][2].
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: tests/framework/utils_imdsv2.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import sys
+import time
+
+import requests
+
+class IMDSv2Client:
+    IMDSV2_HEADER_TOKEN_TTL = "X-aws-ec2-metadata-token-ttl-seconds"
+    IMDSV2_HEADER_TOKEN = "X-aws-ec2-metadata-token"
+
+    def __init__(self, endpoint="http://169.254.169.254/latest"):
```
Comment (mattschlebusch) [inline tests/framework/utils_imdsv2.py:17]: What does this IP address point to? Will it always be available? If this points at a test instance whose address may change, maybe a DNS resource needs to be synthesized that we can route requests to that will always have an instance behind it to process the request?

File: tests/framework/utils_imdsv2.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            raise Exception(f"IMDSv2 returned {res.status_code} for {url}")
+        return res.text
+
+
+IMDS_V2 = IMDSv2Client()
+
+
+import subprocess
+
+IMDSV2_HDR_TOKEN_TTL = "X-aws-ec2-metadata-token-ttl-seconds"
```
Comment (mattschlebusch) [inline tests/framework/utils_imdsv2.py:53]: Is it possible to have these constants defined once globally? (So we can drop the duplicated definition in the `IMDSv2Client` as well)

File: tests/framework/utils_imdsv2.py (hunk 1 above)
Comment (pb8o) [reply tests/framework/utils_imdsv2.py:17]: This is the IMDS address. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html

File: tests/framework/utils_imdsv2.py (hunk 2 above)
Comment (pb8o) [reply tests/framework/utils_imdsv2.py:53]: Yes, I was trying something that wasn't intended to show in this PR. I will clean up before taking the PR out of draft.

File: tests/framework/utils_imdsv2.py (hunk 1 above)
Comment (mattschlebusch) [reply tests/framework/utils_imdsv2.py:17]: Ah I see thanks. might be worth another named constant just for easier reading.

**Review by dianpopa — APPROVED** [review]

File: tests/framework/utils_cpuid.py [hunk 3]
```diff
@@ -46,10 +47,8 @@ def get_cpu_model_name():
 
 
 def get_instance_type():
-    """Get the instance type through IMDS."""
-    imds_cmd = "curl http://169.254.169.254/latest/meta-data/instance-type"
-    _, stdout, _ = run_cmd(imds_cmd)
-    return stdout
+    """Get the instance type through IMDSv2"""
+    return imdsv2_get("/meta-data/instance-type")
```
Comment (mattschlebusch) [inline tests/framework/utils_cpuid.py:51]: Fancy.

**Review by mattschlebusch — APPROVED** [review]
LGTM.

**Review by mattschlebusch — APPROVED** [review]

File: tests/framework/utils_imdsv2.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        headers = {IMDSV2_HDR_TOKEN: self.get_token()}
+        url = f"{self.endpoint}/{self.version}{path}"
+        res = requests.get(url, headers=headers)
+        if res.status_code != 200:
+            raise Exception(f"IMDSv2 returned {res.status_code} for {url}")
+        return res.text
+
+
+IMDS_V2 = IMDSv2Client()
+imdsv2_get = IMDS_V2.get
```
Comment (gc-plp) [inline tests/framework/utils_imdsv2.py:63]: This declaration doesn't seem to be needed.

Using the `IMDS_V2` instance could be done like:

```python

from framework.utils_imdsv2 import IMDS_V2

# using it
print(IMDS_V2.get())

**Review by JonathanWoollett-Light — APPROVED** [review]

**Pull request merged**